### PR TITLE
listing all buildtest config profiles list

### LINKED ISSUE #1490: [FEATURE]: list all profiles
### Please describe your feature

add a command `buildtest config profiles list` to report a list of profiles. The command will display a list of profiles as raw text similar to `buildtest config compilers`. 

We should have an option `--yaml` to print output of all profiles in yaml format. 
We can consider rendering output so let's have `--theme` option so we can use [rich.syntax](https://rich.readthedocs.io/en/stable/syntax.html) to print output in different themes

# TODO

- [x] Add command menu  in  argparse https://github.com/buildtesters/buildtest/blob/devel/buildtest/cli/__init__.py
- [x] Add implementation https://github.com/buildtesters/buildtest/blob/devel/buildtest/cli/config.py
- [x] Add entry in `buildtest help config` https://github.com/buildtesters/buildtest/blob/devel/buildtest/cli/help.py
- [x] Add option in [bash completion](https://github.com/buildtesters/buildtest/blob/devel/bash_completion.sh) script
- [x] Add regression test in https://github.com/buildtesters/buildtest/blob/devel/tests/cli/test_config.py 
- [ ] Add documentation in https://buildtest.readthedocs.io/en/devel/configuring_buildtest.html

### Suggest potential solution

_No response_

### Additional Information

_No response_

### Post question in Slack

- [X] I agree that I posted my question in [slack](https://hpcbuildtest.slack.com/) before creating this issue

### Is there an existing issue

- [X] I confirm there is no existing [issue](https://github.com/buildtesters/buildtest/issues) for this issue

### PR DESCRIPTION
**Pull request by kingxleyy**
add a command buildtest config profiles list

### CODE REVIEW

File: buildtest/cli/config.py [hunk 1]
```diff
@@ -153,6 +155,9 @@ def view_configuration(configuration, theme=None, pager=None):
     console.rule(configuration.file)
     console.print(syntax)
 
+def view_profiles(configuration, theme=None, pager=None):
+    """Display the list of profiles for buildtest configuration file.This implements command ``buildtest config profiles list`
+    """
```
Comment (Mendi03) [inline buildtest/cli/config.py:169]: you have no code for this method.

File: bash_completion.sh [hunk 2]
```diff
@@ -261,7 +261,9 @@ _buildtest ()
           local opts="-h --help"
           COMPREPLY=( $( compgen -W "${opts}" -- $cur ) );;
         view|v)
-          local opts="--help --pager --theme -h -t"
+        COMPREPLY=( $( compgen -W "${opts}" -- $cur ) );;
+        profiles)
+          local opts="--help --pager --theme -h -t --yaml"
```
Comment (Mendi03) [inline bash_completion.sh]: you are modifying the code for `view`. Make sure you leave it as it was.  You did not write `profile` correctly nor put it in the correct lines, this is what is causing errors. Look at the code carefully and write the code syntactically correct.

File: bash_completion.sh [hunk 3]
```diff
@@ -237,7 +237,7 @@ _buildtest ()
     ;;
 
     config|cg)
-      local cmds="-h --help co compilers e edit ex executors p path systems validate v view"
+      local cmds="-h --help co compilers e edit ex executors p path systems validate v view profiles"
```
Comment (Mendi03) [inline bash_completion.sh]: `profiles` should be `profile`. This is what you wrote in buildtest/cli/config.py

File: buildtest/cli/config.py [hunk 4]
```diff
@@ -23,6 +23,9 @@ def config_cmd(args, configuration, editor, system):
     if args.config in ["view", "v"]:
         view_configuration(configuration, theme=args.theme, pager=args.pager)
 
+    if args.config in ["profile"]:
+        view_profiles(configuration, theme=args.theme, pager=args.pager)
+
```
Comment (Mendi03) [inline buildtest/cli/config.py]: The name of the method is not the same as the method definition in line 158. Also, I believe you might need to add the yaml argument here.

File: buildtest/cli/config.py (hunk 1 above)
Comment (Mendi03) [reply buildtest/cli/config.py:169]: If this prints the requested information needed from the feature then it is correct. If not, you need to probably add code for this method. Not sure myself.

File: buildtest/cli/__init__.py [hunk 5]
```diff
@@ -396,7 +398,29 @@ def misc_menu(subparsers):
         ],
         help="Show help message for command",
     )
-
+def profile_menu(subparsers):
+    profile_parser = subparsers.add_parser(
```
Comment (Mendi03) [inline buildtest/cli/__init__.py:402]: Indentation looks incorrect

File: buildtest/cli/config.py (hunk 4 above)
Comment (kingxleyy) [reply buildtest/cli/config.py]: fixed

File: buildtest/cli/__init__.py [hunk 6]
```diff
@@ -396,7 +398,29 @@ def misc_menu(subparsers):
...
+            "executors",
+            "ex",
+            "path",
+            "p",
+            "systems",
+            "validate",
+            "view",
+            "-v"
+        ],
+    )
```
Comment (shahzebsiddiqui) [inline buildtest/cli/__init__.py]: This does not make any sense. You should expect to see a command name `buildtest config profiles` and this is not doing what you want. If you dont see this command then you have not done this correctly.

File: buildtest/cli/config.py [hunk 7]
```diff
@@ -23,6 +23,9 @@ def config_cmd(args, configuration, editor, system):
     if args.config in ["view", "v"]:
         view_configuration(configuration, theme=args.theme, pager=args.pager)
 
+    if args.config in ["profile"]:
+        view_profile(configuration, theme=args.theme, pager=args.pager)
+
```
Comment (shahzebsiddiqui) [inline buildtest/cli/config.py]: change to `elif`

Comment (kingxleyy) [reply buildtest/cli/config.py]: done

File: buildtest/cli/__init__.py [hunk 8]
```diff
@@ -396,6 +398,23 @@ def misc_menu(subparsers):
...
+        "-t",
+        compliers="buildtest config profile",
+        config=["-h"],
+        "help",
+        "--theme",
+        metavar="Color Themes",
+        help="Specify a color theme, Pygments style to use when displaying output. See https://pygments.org/docs/styles/#getting-a-list-of-available-styles for available themese",
+        choices=list(STYLE_MAP.keys()),
+    )
+    return parent_parser
```
Comment (Mendi03) [inline buildtest/cli/__init__.py]: Not sure about the purpose of this method. I don't think you need this in the first place because the `profile` command specified in the issue needs to be under the `config_menu` method.

File: buildtest/cli/compilers.py
Comment (Mendi03) [inline buildtest/cli/compilers.py:1]: according to the issue, I don't think you needed to modify this file whatsoever.

File: buildtest/cli/config.py [hunk 9]
```diff
@@ -23,6 +23,9 @@ def config_cmd(args, configuration, editor, system):
     if args.config in ["view", "v"]:
         view_configuration(configuration, theme=args.theme, pager=args.pager)
 
+    elif args.config in ["profile"]:
+        view_profile(configuration, theme=args.theme, pager=args.pager)
+
```
Comment (Mendi03) [inline buildtest/cli/config.py]: Here it is saying that the `view_profile` method should be called when someone writes the `profile` argument in the terminal. However, you have not written any of this logic in the __init__.py file. As I stated on a previous comment, you need to add this somewhere in the config_menu method.

File: buildtest/cli/__init__.py [hunk 10]
```diff
@@ -975,7 +994,9 @@ def config_menu(subparsers, parent_parser):
     compilers = subparsers_config.add_parser(
         "compilers", aliases=["co"], help="Search compilers"
     )
-
+    profile = subparsers_config.add_parser(
+        "profile", aliases=["profile"], help="Search profile"
+    )
```
Comment (Mendi03) [inline buildtest/cli/__init__.py]: This looks right, you might need to add an argument `list` to this subparser as specified in the issue though.

Comment (kingxleyy) [reply buildtest/cli/__init__.py]: done

File: buildtest/cli/__init__.py [hunk 11]
```diff
@@ -396,6 +398,23 @@ def misc_menu(subparsers):
...
     )
+def profile_menu(subparsers):
+
+    config_profiles = subparsers_buildpsec.add_parser(
+        parents=[parent_parser["theme"]],
+    )
+    parent_parser["theme"] = argparse.ArgumentParser(add_help=False)
+    parent_parser["theme"].add_argument(
+        "-t",
+        compliers="buildtest config profile",
```
Comment (shahzebsiddiqui) [inline buildtest/cli/__init__.py]: why do you have this line, please remove `compliers="buildtest config profile",`

File: buildtest/cli/__init__.py [hunk 12]
```diff
@@ -1063,6 +1084,10 @@ def config_menu(subparsers, parent_parser):
         "compiler_names", nargs="*", help="Specify compiler name to test"
     )
 
+    # buildtest config profile
+    profile.add_argument(
+        "-y", "--yaml", "--theme","view", help="List Profile details in YAML Format"
```
Comment (Mendi03) [inline buildtest/cli/__init__.py]: This is not right, you need a `list` argument, not a list of arguments.

File: buildtest/cli/__init__.py [hunk 13]
```diff
@@ -1063,6 +1084,10 @@ def config_menu(subparsers, parent_parser):
...
+    )
```
Comment (shahzebsiddiqui) [inline buildtest/cli/__init__.py]: this is not correct. You can't have `--theme` and `view`.

File: buildtest/cli/__init__.py (hunk 11 above)
Comment (kingxleyy) [reply buildtest/cli/__init__.py]: done

File: buildtest/cli/config.py [hunk 14]
```diff
@@ -153,6 +155,11 @@ def view_configuration(configuration, theme=None, pager=None):
     console.rule(configuration.file)
     console.print(syntax)
 
+def view_profile(configuration, theme=None, pager=None):
+    """Display the list of profile for buildtest configuration file.This implements command ``buildtest config profiles list`
+    """
+
+    console.print(configuration.file)
```
Comment (shahzebsiddiqui) [inline buildtest/cli/config.py]: this is not right for the implementation. You need to index the dictionary and list all the profiles. the `configuration.file` is the full path to configuration file.

File: buildtest/settings/config.yml [hunk 15]
```diff
@@ -1,45 +1,23 @@
 system:
   generic:
-    # specify list of hostnames where buildtest can run for given system record
-    hostnames: [".*"]
-
-    # system description
```
Comment (shahzebsiddiqui) [inline buildtest/settings/config.yml]: @kingxleyy this was probably my mistake. Please revert this change back to original file shown in `devel` branch. Just copy the entire file and commit it locally and push it so there are no changes shown for this file

Comment (kingxleyy) [reply buildtest/settings/config.yml]: done

File: bash_completion.sh [hunk 16]
```diff
@@ -269,6 +272,19 @@ _buildtest ()
...
         ;;
+        profile)
+          local opts="--help -h list"
+          COMPREPLY=( $( compgen -W "${opts}" -- $cur ) )
+          if [[ "${prev}" == "list" ]]; then
+            local opts="--help  --theme --yaml -h -t -y"
+            COMPREPLY=( $( compgen -W "${opts}" -- $cur ) )
+          fi
+
+          case "${prev}" in --theme|-t)
```
Comment (shahzebsiddiqui) [inline bash_completion.sh]: can you remove the `-t` option this will be removed in https://github.com/buildtesters/buildtest/pull/1576

File: bash_completion.sh [hunk 17]
```diff
@@ -269,6 +272,19 @@ _buildtest ()
             return
           esac
         ;;
+        profile)
+          local opts="--help -h list"
+          COMPREPLY=( $( compgen -W "${opts}" -- $cur ) )
+          if [[ "${prev}" == "list" ]]; then
+            local opts="--help  --theme --yaml -h -t -y"
```
Comment (shahzebsiddiqui) [inline bash_completion.sh]: same here remove the `-t`

File: buildtest/cli/config.py [hunk 18]
```diff
@@ -154,6 +159,31 @@ def view_configuration(configuration, theme=None, pager=None):
     console.print(syntax)
 
 
+def list_profiles(configuration, theme=None, print_yaml=None):
+    """Display the list of profile for buildtest configuration file.This implements command ``buildtest config profiles list``
```
Comment (shahzebsiddiqui) [inline buildtest/cli/config.py]: add a space after the period `file. This`

File: buildtest/settings/config.yml (hunk 15 above)
Comment (shahzebsiddiqui) [inline buildtest/settings/config.yml]: please dont commit `buildtest/settings/config.yml` this was a result of running the regression test which autoupdates the file. Don't do `git add .` or something. Just copy the content from `devel` branch and paste it in that should avoid showing this commit diff for this file

File: buildtest/cli/__init__.py [hunk 19]
```diff
@@ -988,6 +988,24 @@ def config_menu(subparsers, parent_parser):
     compilers = subparsers_config.add_parser(
         "compilers", aliases=["co"], help="Search compilers"
     )
+    # buildtest config profile
+    profile = subparsers_config.add_parser(
+        "profile", help="Query profile from buildtest configuration"
```
Comment (shahzebsiddiqui) [inline buildtest/cli/__init__.py]: can you change the command to `profiles` with an `s`. So do the following

```python
    profile = subparsers_config.add_parser(
        "profiles", help="Query profile from buildtest configuration")
```

Note the `s` is for the argument in add_parser not the variable `profile`.

File: buildtest/cli/__init__.py [hunk 20]
```diff
@@ -988,6 +988,24 @@ def config_menu(subparsers, parent_parser):
...
+    )
+    subparsers_profile = profile.add_subparsers(
+        description="Query information about buildtest profiles",
+        dest="profile",
```
Comment (shahzebsiddiqui) [inline buildtest/cli/__init__.py]: for consistency make this `profiles`.

File: buildtest/cli/config.py [hunk 21]
```diff
@@ -20,9 +20,14 @@ def config_cmd(args, configuration, editor, system):
         configuration (buildtest.config.SiteConfiguration): An instance of SiteConfiguration class
         system (buildtest.system.BuildTestSystem): An instance of BuildTestSystem class
     """
+
     if args.config in ["view", "v"]:
         view_configuration(configuration, theme=args.theme, pager=args.pager)
 
+    elif args.config in ["profile"]:
```
Comment (shahzebsiddiqui) [inline buildtest/cli/config.py]: same here change this to `profiles`

File: buildtest/cli/config.py [hunk 22]
```diff
@@ -20,9 +20,14 @@ def config_cmd(args, configuration, editor, system):
...
+        if args.profile == "list":
```
Comment (shahzebsiddiqui) [inline buildtest/cli/config.py]: and here to change to 

```python
if args.profiles == "list":
```

File: bash_completion.sh [hunk 23]
```diff
@@ -237,7 +237,10 @@ _buildtest ()
     ;;
 
     config|cg)
-      local cmds="-h --help co compilers e edit ex executors p path systems val validate v view"
+
+
+      local cmds="-h --help co compilers e edit ex executors p path profile systems validate v view"
```
Comment (shahzebsiddiqui) [inline bash_completion.sh]: change `profile` -> `profiles`

File: bash_completion.sh [hunk 24]
```diff
@@ -269,6 +272,19 @@ _buildtest ()
             return
           esac
         ;;
+        profile)
```
Comment (shahzebsiddiqui) [inline bash_completion.sh]: change this to `profiles`

Comment (kingxleyy) [reply bash_completion.sh]: done

File: bash_completion.sh (hunk 23 above)
Comment (kingxleyy) [reply bash_completion.sh]: done

File: buildtest/cli/config.py (hunk 22 above)
Comment (kingxleyy) [reply buildtest/cli/config.py]: done

File: buildtest/cli/config.py (hunk 21 above)
Comment (kingxleyy) [reply buildtest/cli/config.py]: done

File: buildtest/cli/__init__.py (hunk 20 above)
Comment (kingxleyy) [reply buildtest/cli/__init__.py]: done

File: buildtest/cli/__init__.py (hunk 19 above)
Comment (kingxleyy) [reply buildtest/cli/__init__.py]: done

File: buildtest/settings/config.yml (hunk 15 above)
Comment (kingxleyy) [reply buildtest/settings/config.yml]: noted

File: buildtest/cli/config.py (hunk 18 above)
Comment (kingxleyy) [reply buildtest/cli/config.py]: done

File: bash_completion.sh (hunk 17 above)
Comment (kingxleyy) [reply bash_completion.sh]: done

File: buildtest/settings/config.yml (hunk 15 above)
Comment (shahzebsiddiqui) [reply buildtest/settings/config.yml]: @kingxleyy i still see this file ``buildtest/settings/config.yml` in commit diff
